File: backend/app/glitchtip.py
import os
import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from fastapi import Request
import logging

def init_glitchtip():
    """Initialize GlitchTip (Sentry) for error tracking."""
    
    # Check if DSN is configured
    dsn = os.getenv("GLITCHTIP_DSN")
    if not dsn:
        logging.warning("GlitchTip DSN not configured - error tracking disabled")
        return
    
    # Initialize Sentry SDK
    sentry_sdk.init(
        dsn=dsn,
        
        # Set environment
        environment=os.getenv("ENVIRONMENT", "production"),
        
        # Performance monitoring (lower rate for production)
        traces_sample_rate=0.1,
        
        # Integrations
        integrations=[
            FastApiIntegration(auto_enabling_integrations=False),
            SqlalchemyIntegration(),
        ],
        
        # Error filtering
        before_send=filter_errors,
        
        # Additional configuration
        attach_stacktrace=True,
        send_default_pii=False,  # Don't send personally identifiable info
    )
    
    logging.info("GlitchTip error tracking initialized")

def filter_errors(event, hint):
    """Filter out errors we don't want to track."""
    
    # Don't track in development
    if os.getenv("ENVIRONMENT") == "development":
        return None
    
    # Get the exception
    if 'exc_info' in hint:
        exc_type, exc_value, tb = hint['exc_info']
        
        # Skip common HTTP client errors
        if exc_type.__name__ in [
            'HTTPException',
            'ValidationError',
            'RequestValidationError'
        ]:
            # Only track 5xx errors, not 4xx client errors
            if hasattr(exc_value, 'status_code'):
                if 400 <= exc_value.status_code < 500:
                    return None
        
        # Skip database connection errors that are temporary
        error_message = str(exc_value).lower()
        if any(msg in error_message for msg in [
            'connection refused',
            'connection timeout',
            'connection reset',
            'connection closed'
        ]):
            return None
    
    return event

def capture_exception(error: Exception, context: dict = None):
    """Manually capture an exception with optional context."""
    with sentry_sdk.configure_scope() as scope:
        if context:
            for key, value in context.items():
                scope.set_context(key, value)
        sentry_sdk.capture_exception(error)

def capture_message(message: str, level: str = "info", context: dict = None):
    """Capture a message with optional context."""
    with sentry_sdk.configure_scope() as scope:
        if context:
            for key, value in context.items():
                scope.set_context(key, value)
        sentry_sdk.capture_message(message, level=level)

def set_user_context(user_id: str, email: str = None, extra: dict = None):
    """Set user context for error tracking."""
    user_data = {"id": user_id}
    if email:
        user_data["email"] = email
    if extra:
        user_data.update(extra)
    
    sentry_sdk.set_user(user_data)

def add_breadcrumb(message: str, category: str = "custom", data: dict = None):
    """Add a breadcrumb for debugging context."""
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        data=data or {},
    )

# Middleware to add request context
def add_request_context(request: Request):
    """Add request context to error tracking."""
    with sentry_sdk.configure_scope() as scope:
        scope.set_context("request", {
            "url": str(request.url),
            "method": request.method,
            "headers": dict(request.headers),
            "query_params": dict(request.query_params),
        })

# Custom exception handler for better error tracking
async def glitchtip_exception_handler(request: Request, exc: Exception):
    """Custom exception handler that captures errors to GlitchTip."""
    
    # Add request context
    add_request_context(request)
    
    # Capture the exception
    capture_exception(exc, {
        "request_info": {
            "url": str(request.url),
            "method": request.method,
            "user_agent": request.headers.get("user-agent"),
        }
    })
    
    # Re-raise the exception to let FastAPI handle it
    raise exc

File: frontend/src/lib/glitchtip.ts
'use client';

import React from 'react';
import * as Sentry from '@sentry/nextjs';

// Initialize Sentry (GlitchTip is Sentry-compatible)
export function initGlitchTip() {
  if (!process.env.NEXT_PUBLIC_GLITCHTIP_DSN) {
    console.warn('GlitchTip DSN not configured');
    return;
  }

  Sentry.init({
    dsn: process.env.NEXT_PUBLIC_GLITCHTIP_DSN,
    
    // Set environment
    environment: process.env.NODE_ENV || 'development',
    
    // Performance monitoring
    tracesSampleRate: process.env.NODE_ENV === 'production' ? 0.1 : 1.0,
    
    // Session replay (optional, uses more resources)
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,
    
    // Error filtering
    beforeSend(event, hint) {
      // Don't send errors in development
      if (process.env.NODE_ENV === 'development') {
        console.error('Error captured by GlitchTip:', event);
        return null;
      }
      
      // Filter out common non-critical errors
      const error = hint.originalException;
      if (error && typeof error === 'object') {
        const message = 'message' in error ? error.message : '';
        
        // Skip network errors that are likely user issues
        if (typeof message === 'string') {
          if (message.includes('Network Error') || 
              message.includes('Failed to fetch') ||
              message.includes('ERR_NETWORK') ||
              message.includes('ERR_INTERNET_DISCONNECTED')) {
            return null;
          }
        }
      }
      
      return event;
    },
    
    // Additional options
    integrations: [
      // Add performance monitoring
      Sentry.browserTracingIntegration(),
      // Add session replay (optional)
      Sentry.replayIntegration(),
    ],
  });
}

// Helper functions for manual error reporting
export function captureError(error: Error, context?: Record<string, any>) {
  Sentry.withScope((scope) => {
    if (context) {
      Object.keys(context).forEach(key => {
        scope.setContext(key, context[key]);
      });
    }
    Sentry.captureException(error);
  });
}

export function captureMessage(message: string, level: 'error' | 'warning' | 'info' = 'info') {
  Sentry.captureMessage(message, level);
}

export function setUserContext(user: { id: string; email?: string; [key: string]: any }) {
  Sentry.setUser(user);
}

export function addBreadcrumb(message: string, category: string = 'custom', data?: Record<string, any>) {
  Sentry.addBreadcrumb({
    message,
    category,
    data,
    timestamp: Date.now() / 1000,
  });
}

// Error boundary component (simplified to avoid TypeScript issues)
export function withErrorBoundary<T extends object>(
  Component: React.ComponentType<T>
) {
  return Sentry.withErrorBoundary(Component, {
    fallback: ({ error, resetError }) => 
      React.createElement('div', { 
        className: "min-h-screen flex items-center justify-center bg-gray-50" 
      },
        React.createElement('div', { className: "text-center" },
          React.createElement('h2', { 
            className: "text-xl font-bold text-gray-900 mb-4" 
          }, "Something went wrong"),
          React.createElement('p', { 
            className: "text-gray-600 mb-4" 
          }, "An error occurred while loading this page."),
          React.createElement('button', {
            onClick: resetError,
            className: "px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600"
          }, "Try again")
        )
      ),
    beforeCapture: (scope, error) => {
      scope.setTag('errorBoundary', true);
    },
  });
}